ubi_manifest/app/utils.py

```diff
@@ -0,0 +1,112 @@
+    if ubi_repos:
+        all_items.extend(get_ubi_items(ubi_repos, app_conf))
+
+    _LOG.info("Determined items for depsolving: %s", all_items)
+    return all_items
+
+
+def get_client_tools_items(
+    provided_repo_ids: list[str], config_url: str
+) -> list[dict[str, Any]]:
+    """
+    Returns a list of items for depsolving of client-tools repos.
+    For these we don't use full depsolving so it is performed only
+    on the provided repo and no group is needed to be determined.
+    """
+    items: list[dict[str, Any]] = []
+    for repo in provided_repo_ids:
+        items.append({"repo_group": [repo], "url": config_url})
+    return items
+
+
+def get_ubi_items(provided_repo_ids: list[str], app_conf: Any) -> list[dict[str, Any]]:
+    """
+    Returns a list of items for depsolving of ubi repos.
+    """
+    items: list[dict[str, Any]] = []
+    needed_repo_groups: dict[str, set[str]] = {}
+
+    config_url = app_conf.content_config["ubi"]
+    allowed_ubi_repo_groups = get_allowed_ubi_repo_groups(config_url, app_conf)
+
+    # Compare provided repo_ids with the generated allowed repo groups and pick
+    # those needed for this request
+    for repo_id in provided_repo_ids:
+        for group_key, repo_group in allowed_ubi_repo_groups.items():
+            if repo_id in repo_group:
+                needed_repo_groups.setdefault(group_key, repo_group)
+                break
+
+    # Create items for depsolving
+    for repo_group in needed_repo_groups.values():
+        items.append({"repo_group": list(sorted(repo_group)), "url": config_url})
+
+    return items
+
+
+def get_allowed_ubi_repo_groups(config_url: str, app_conf: Any) -> dict[str, set[str]]:
+    """
+    Determines the allowed repo groups based on the the available ubi config and Pulp data.
+    """
+    allowed_ubi_repo_groups: dict[str, set[str]] = {}
+
+    _LOG.info("Loading config from %s", config_url)
+
+    loader = ubiconfig.get_loader(config_url)
+    configs = loader.load_all()
+    # Use only configs for major versions
+    configs = [conf for conf in configs if "." not in conf.version]
+
+    with make_pulp_client(app_conf) as client:
+        for conf in configs:
+            repos = get_repo_ids_from_cs(client, conf.content_sets.rpm.output)
+            for repo in repos:
+                # Create groups by version and architecture
+                key = f"ubi{conf.version}-{repo.arch}"
+                allowed_ubi_repo_groups.setdefault(key, set()).add(repo.id)
+
+        return allowed_ubi_repo_groups
+
+
+def get_repo_ids_from_cs(client: Client, ubi_binary_cs: str) -> Any:
+    """
+    Returns a list of YumRepositories associated with the given content set
+    and which have the ubi_population note set to True.
+    """
+    repos = client.search_repository(
+        Criteria.and_(
+            Criteria.with_field("notes.content_set", ubi_binary_cs),
+            Criteria.with_field("ubi_population", True),
+        )
+    )
+    return repos.result()
```

Comment:
You don't have to call `result()` here, returning just `repos` should work just fine in `get_allowed_ubi_repo_groups`  function.